File: LeetCode/LeetCode1_50/5_Longest_Palindromic_Substring.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LeetCode
{
    public static partial class LeetCode1_50
    {
        public static string LongestPalindrome(string s)
        {
            if (string.IsNullOrEmpty(s))
            {
                return string.Empty;
            }

            int start = 0, len = 0;

            for (int i = 0; i < s.Length; i++)
            {
                int lenOdd = CheckExpand(i, i);
                int lenEven = CheckExpand(i, i + 1);
                int maxLen = Math.Max(lenOdd, lenEven);
                if (maxLen > len)
                {
                    start = i - (maxLen - 1) / 2;
                    len = maxLen;
                }

            }
            return s.Substring(start, len);

            int CheckExpand(int left, int right)
            {
                while (left >= 0 && right < s.Length && s[left] == s[right])
                {
                    left--;
                    right++;
                }
                return right - left - 1;
            }
        }
    }
}

File: LeetCode/LeetCode1_50/5_Longest_Palindromic_Substring.cpp
#include <vector>
#include <string>

#include "../..//LeetCodeCpp/LeetCode1_50.h"

using namespace std;

string LeetCode::LeetCode1_50::longestPalindrome(string s)
{
    if (s.size() < 2) return s;
    int pos = 0, len = 0, i = 0;
    while (i < s.size())
    {
        int l = i, r = i;
        while (r < s.size() - 1 && s[r + 1] == s[r]) ++r;
        i = r + 1;
        while (r < s.size() - 1 && l > 0 && s[r + 1] == s[l - 1]) { ++r; --l; }
        int nlen = r - l + 1;
        if (nlen > len) { pos = l; len = nlen; }
    }
    return s.substr(pos, len);
}